app/src/forms/rbac/service.js

```diff
@@ -0,0 +1,440 @@
+const { FormRoleUser, User, UserFormRoles, UserFormPermissions } = require('../common/models');
+
+const {transaction} = require('objection');
+const {v4: uuidv4} = require('uuid');
+
+const toUserForms = (userPerms, userRoles) => {
+  if (!(userPerms || userRoles) && !(userPerms.length || userRoles.length)) {
+    return null;
+  }
+
+  function toUser(item) {
+    return {
+      id: item.id,
+      keycloakId: item.keycloakId,
+      username: item.username,
+      fullName: item.fullName,
+      email: item.email,
+      firstName: item.firstName,
+      lastName: item.lastName,
+      forms: []
+    };
+  }
+
+  function toForm(item) {
+    return {
+      id: item.formId,
+      name: item.formName,
+      public: item.public,
+      active: item.active,
+      formVersionId: item.formVersionId,
+      version: item.version,
+      roles: [],
+      permissions: []
+    };
+  }
+
+  const result = [];
+
+  if (userPerms && userPerms.length) {
+    userPerms.forEach(item => {
+      let user = result.find(u => u.id === item.id);
+      if (!user) {
+        user = toUser(item);
+        result.push(user);
+      }
+      let form = user.forms.find(f => f.id === item.formId);
+      if (!form) {
+        form = toForm(item);
+        user.forms.push(form);
+      }
+      form.permissions = (item.permissions) ? item.permissions.sort() : [];
+    });
+  }
+
+  if (userRoles && userRoles.length) {
+    userRoles.forEach(item => {
+      let user = result.find(u => u.id === item.id);
+      if (!user) {
+        user = toUser(item);
+        result.push(user);
+      }
+      let form = user.forms.find(f => f.id === item.formId);
+      if (!form) {
+        form = toForm(item);
+        user.forms.push(form);
+      }
+      form.roles = (item.roles) ? item.roles.sort() : [];
+    });
+  }
+
+  return result;
+};
+
+const toFormUsers = (userPerms, userRoles) => {
+  if (!(userPerms || userRoles) && !(userPerms.length || userRoles.length)) {
+    return null;
+  }
+
+  function toForm(item) {
+    return {
+      id: item.formId,
+      name: item.formName,
+      public: item.public,
+      active: item.active,
+      formVersionId: item.formVersionId,
+      version: item.version,
+      users: []
+    };
+  }
+  function toUser(item) {
+    return {
+      id: item.id,
+      keycloakId: item.keycloakId,
+      username: item.username,
+      fullName: item.fullName,
+      email: item.email,
+      firstName: item.firstName,
+      lastName: item.lastName,
+      roles: [],
+      permissions: []
+    };
+  }
```

Comment:
These helper transform functions appear duplicated at the top of the file as well. As these are generally more "schema" transform functions, it may not be a bad idea to create a separate transform.js supplementary file, or even just a top level const transformer that can be module.exported where its sole purpose is to just do an object to object mapping. Also makes it easier to unit test this as a sole unit later instead of having it be defined and embedded down a level, where we'd only be able to test it implicitly.